Название проекта: Исследование надёжности заёмщиков

Заказчик: Кредитный отдел банка

Цели проекта: 
- Установить влияние семейного положения и количества детей клиента банка на факт погашения кредита в срок
- Построить модель для предсказания надежности заемщика


In [1]:
import pandas as pd

### 1. Открытие файла и изучение данных. 

In [2]:
bank_data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

Получим информацию о количестве и типе данных:

Просмотри и оценим данные визуально:

In [3]:
bank_data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [4]:
bank_data.tail(10)

children  days_employed  dob_years       education  education_id  \
21515         1    -467.685130         28         среднее             1   
21516         0    -914.391429         42          высшее             0   
21517         0    -404.679034         42          высшее             0   
21518         0  373995.710838         59         СРЕДНЕЕ             1   
21519         1   -2351.431934         37  ученая степень             4   
21520         1   -4529.316663         43         среднее             1   
21521         0  343937.404131         67         среднее             1   
21522         1   -2113.346888         38         среднее             1   
21523         3   -3112.481705         38         среднее             1   
21524         2   -1984.507589         40         среднее             1   

          family_status  family_status_id gender income_type  debt  \
21515   женат / замужем                 0      F   сотрудник     1   
21516   женат / замужем                 0      F   компаньон     0   
21517  гражданский брак                 1      F   компаньон     0   
21518   женат / замужем                 0      F   пенсионер     0   
21519         в разводе                 3      M   сотрудник     0   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                            purpose  
21515  109486.327999              заняться образованием  
21516  322807.776603               покупка своего жилья  
21517  178059.553491       на покупку своего автомобиля  
21518  153864.650328               сделка с автомобилем  
21519  115949.039788  покупка коммерческой недвижимости  
21520  224791.862382                  операции с жильем  
21521  155999.806512               сделка с автомобилем  
21522   89672.561153                       недвижимость  
21523  244093.050500       на покупку своего автомобиля  
21524   82047.418899              на покупку автомобиля

In [5]:
bank_data.sample(10, random_state = 42)

children  days_employed  dob_years education  education_id  \
423           0    -191.167760         56   среднее             1   
3522          0   -2319.817259         58   среднее             1   
8760          1   -2990.578297         34    высшее             0   
20695         0    -926.209452         28   среднее             1   
4351          0   -2524.302106         42   СРЕДНЕЕ             1   
3002          0            NaN         28    высшее             0   
17306         0   -5000.830630          0   среднее             1   
2830          0    -574.995923         40   среднее             1   
14822         2            NaN         33   среднее             1   
10347         1    -713.140231         48   среднее             1   

               family_status  family_status_id gender income_type  debt  \
423          женат / замужем                 0      M   сотрудник     0   
3522   Не женат / не замужем                 4      F   сотрудник     0   
8760        гражданский брак                 1      M   компаньон     0   
20695  Не женат / не замужем                 4      F   сотрудник     0   
4351         женат / замужем                 0      F   сотрудник     0   
3002         женат / замужем                 0      M   компаньон     0   
17306       гражданский брак                 1      F   компаньон     0   
2830         женат / замужем                 0      F   сотрудник     0   
14822        женат / замужем                 0      F   сотрудник     0   
10347        женат / замужем                 0      M   компаньон     0   

        total_income                        purpose  
423    138653.748793                     автомобили  
3522   123152.627177                  покупка жилья  
8760   232380.737167                        свадьба  
20695  210617.086601  получение высшего образования  
4351    78487.540219           сделка с автомобилем  
3002             NaN     покупка жилой недвижимости  
17306   83019.106740                        свадьба  
2830   161158.806308                     автомобили  
14822            NaN          заняться образованием  
10347  177474.885995                   ремонт жилью

In [6]:
bank_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Посмотрев на данные можем отметить наличие пропусков в данных, отрицательных значений, в параметрах, где не должно быть значений меньше 0, а так же написание одного значения признака в различных регистрах, что будет мешать анализу.

### 2. Предобработка данных

#### 2.1.Заполнение пропущенных значений:

In [7]:
bank_data.isnull().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Видим одинаковое количество пропусков для двух признаков. Проверим соотвествуют ли эти пропуски одним объектам или разным.


In [8]:
bank_data[(bank_data['days_employed'].isnull()==True)&(bank_data['total_income'].isnull()==True)].count()

children            2174
days_employed          0
dob_years           2174
education           2174
education_id        2174
family_status       2174
family_status_id    2174
gender              2174
income_type         2174
debt                2174
total_income           0
purpose             2174
dtype: int64

**Вывод:**

Ячейки с пропусками: общий трудовой стаж в днях *days_employed* и ежемесячный доход *total_income* относятся к одним и тем же клиентам.

Далее заменим пропуски в столбце *total_income* медианным значениями в зависимости от типа занятости *income_type*.
Пропуски в столбце *days_employed* заменим медианным значением в зависисмости от возраста человека.

Для этого значения общего трудового стажа в днях *days_employed* возьмем по модулю. Так как количество дней не может быть отрицательным, а анализируя зависимость трудовых дней от возраста мы можем сделать вывод, что -(минус) - это опечатка. 

Для удобства работы с данными в дальнейшем - получим словари. 

Получим среднее значение трудового стажа в зависимости от возраста клиента. Возраст - ключ словаря, количество дней - значение по данному ключу.

Получим среднее значение зароботка в зависимости от типа занятости. Тип занятости - ключ словаря, доход - значение по данному ключу.

##### 2.1.2 Замена пропущенных значений

In [9]:
bank_data['days_employed'] = abs(bank_data['days_employed'])
mean_days_dict = bank_data.groupby('dob_years')['days_employed'].median().to_dict()
mean_total_income_dict = bank_data.groupby('income_type')['total_income'].median().to_dict()

In [10]:
#Заполнение пропусков в столбцах days_employed и total_income
unique_income_type = bank_data['income_type'].unique()
unique_days_employed = bank_data['dob_years'].unique()

for inc_type in unique_income_type:
    bank_data.loc[bank_data['total_income'] == 0,'total_income'] =mean_total_income_dict[inc_type]
    
for day in unique_days_employed:
    bank_data.loc[bank_data['days_employed'].isnull()== True,'days_employed'] = mean_days_dict[day]
    

Контроль замены пропусков:

In [11]:
check_income = bank_data[bank_data['total_income'].isnull()==True]['total_income'].count()
check_days = bank_data[bank_data['days_employed'].isnull()==True]['days_employed'].count()
print(f'Количество пропущенных значений в столбце days_employed: {check_days}')
print(f'Количество пропущенных значений в столбце total_income: {check_income}')

Количество пропущенных значений в столбце days_employed: 0
Количество пропущенных значений в столбце total_income: 0


##### 2.1.4 Замена значений данных о количество детей

При проверке данных о количестве детей(children) найдены 2 значения, выделющиеся из списка значений. Первое из них равно -1(минус 1) - количество детей не может быть меньше 0. Второе значение из списка, требующее внимания - 20(двадцать). Такое значение, в принципе возможно, но о том, что это скорее всего ошибка, указывает тот факт, что отстутвуют значения количества детей в инервале от 6 до 19 включительно. Далее изучим строки содержащие данные значения.

Никаких визуальных зависимостей между клиентами, со значением ячейки количество детей равным -1 и 20 визуально не заметно. Можно предположить, что значение 20 получалось путем опечатки человека, который вносил данные, т.к. цифры 2 и 0 находятся рядом на дополнительном циферблате клавиатуры. А значение -1 могло получиться в случае отсутсвия данных о количестве детей.

**Вывод:**
Заменим значения -1, 20 на 0 - как медианное значение.

In [12]:
bank_data.children.value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

In [13]:
bank_data.loc[((bank_data['children']==-1 )|(bank_data['children']==20)),'children']=0

##### 2.2 Приведение вещественных значений к целочисленным

При первоначлаьном анализе типов данных было заметно, что часть данных имеет вещественный тип:

In [14]:
bank_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


2 колонки с данными имеют вещественный тип данных float64 - это колонки *days_employed* и *total_income*. Для удобства работы заменяем значения в этих столбцах на целочисленные и проверим резульат:

In [15]:
bank_data['days_employed'] = bank_data['days_employed'].astype('int64')
bank_data['total_income'] = bank_data['total_income'].fillna(0)
bank_data['total_income'] = bank_data['total_income'].astype('int64')
bank_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   children          21525 non-null  int64 
 1   days_employed     21525 non-null  int64 
 2   dob_years         21525 non-null  int64 
 3   education         21525 non-null  object
 4   education_id      21525 non-null  int64 
 5   family_status     21525 non-null  object
 6   family_status_id  21525 non-null  int64 
 7   gender            21525 non-null  object
 8   income_type       21525 non-null  object
 9   debt              21525 non-null  int64 
 10  total_income      21525 non-null  int64 
 11  purpose           21525 non-null  object
dtypes: int64(7), object(5)
memory usage: 2.0+ MB


**Вывод:**

Все вещественные типы в датасете заменены на целочисленные

#### 2.3 Поиск дубликатов и удаление дубликатов

Возможной причиной появления дубликатов может быть некорректная работы системы банка во время внесения данных.

Перед проверкой дубликатов произведем проверку регистра в данных:

In [16]:
bank_data.education.value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64

Видим, что значения *education* записано в разном регистре, что помешает удалить все дубликаты. Переведем все значения этого столбца в одинаковый регистр и проверим полученные данные:

In [17]:
def My_capitaize(row):
    """
    Функция делает первую букву в верхний регистр, а остальные в нижний
    """
    return row.capitalize()

bank_data['education'] =bank_data['education'].apply(My_capitaize)
bank_data.education.value_counts()

Среднее                15233
Высшее                  5260
Неоконченное высшее      744
Начальное                282
Ученая степень             6
Name: education, dtype: int64

Очистим дубликаты и проверим данные:

In [18]:
print(f'Количество повторений до очистки от дубликатов: {bank_data.duplicated().sum()}. Всего строк {bank_data.days_employed.count()}')

Количество повторений до очистки от дубликатов: 71. Всего строк 21525


In [19]:
bank_data = bank_data.drop_duplicates()
print(f'Количество повторений до очистки от дубликатов: {bank_data.duplicated().sum()}. Всего строк {bank_data.days_employed.count()}')

Количество повторений до очистки от дубликатов: 0. Всего строк 21454


**Вывод:**

Во время проверки данных на наличие дубликатов было обнаружено 71 одно повторение, при размере датафрейма 21525 строка. После очистки от дупликатов размер датафрейма стал 21454.

#### 2.4 Выделение лемм в целях получения кредита

In [20]:
!pip install pymystem3

In [21]:
from pymystem3 import Mystem
from collections import Counter
import nltk
from nltk.stem import SnowballStemmer 
russian_stemmer = SnowballStemmer('russian')

Произведем лемматизацию значений колонки *Purpose* для выявления часто используемых слов:

In [22]:
m = Mystem()
bank_data.purpose.to_list()
string_lemma = " ".join(bank_data.purpose)

In [23]:
dict(Counter(m.lemmatize(string_lemma)))

{'покупка': 5897,
 ' ': 55023,
 'жилье': 4460,
 'приобретение': 461,
 'автомобиль': 4306,
 'дополнительный': 906,
 'образование': 4013,
 'сыграть': 765,
 'свадьба': 2324,
 'операция': 2604,
 'с': 2918,
 'на': 2222,
 'проведение': 768,
 'для': 1289,
 'семья': 638,
 'недвижимость': 6351,
 'коммерческий': 1311,
 'жилой': 1230,
 'строительство': 1878,
 'собственный': 635,
 'подержать': 853,
 'свой': 2230,
 'со': 627,
 'заниматься': 904,
 'сделка': 941,
 'получение': 1314,
 'высокий': 1374,
 'подержанный': 111,
 'профильный': 436,
 'сдача': 651,
 'ремонт': 607,
 '\n': 1}

Вывод:

Из полученных данных выделим уникальные чаще встречающиеся значений. Такие как "покупка", "приобретение", "операция", "с", "на" встречаются часто, но они не позволяют определить категорию к которой отнести данные. Ниже в списке *lemmas* приведены значения лемм для колонки *purpose*. 

In [24]:
lemmas = ['жилье','автомобиль','образование','недвижимость','свадьба']

Для дальнейшей работы произведем стемминг полученных значений лемм. И запишем их в словарь где ключом является результат стемминга леммы, а значением сама лемма.

In [25]:
lemmas_dict = {}
for word in lemmas:
    stemmed_word = russian_stemmer.stem(word)
    lemmas_dict[stemmed_word] = word
lemmas_dict

{'жил': 'жилье',
 'автомобил': 'автомобиль',
 'образован': 'образование',
 'недвижим': 'недвижимость',
 'свадьб': 'свадьба'}

Создадим и вызовем функцию для леммы категории цели получения кредита.

In [26]:
russian_stemmer = SnowballStemmer('russian')
def lem(row):
    row_list = row.split(" ") 
    i = 0
    while i < len(row_list):     
        stemmed_word = russian_stemmer.stem(row_list[i])
        if  stemmed_word in lemmas_dict.keys():
            if stemmed_word == 'жил' or stemmed_word == 'недвижим':
                return 'недвижимость'
            else:
                return lemmas_dict[stemmed_word]
        i+=1
bank_data.loc[:,'lemmatized_purpose'] = bank_data.purpose.apply(lem)

#### 2.5 Категоризация данных

Произведем категоризацию данных:

##### 2.5.1 Отнесем человека к группе в зависимости от заработка:

In [27]:
inocome_quantile = bank_data.total_income.quantile(q=[0.25,0.5,0.75]).to_dict()

def set_income_level(row):
    if int(row['total_income']) <= inocome_quantile[0.25]:
        return 'низкий'
    elif int(row['total_income']) > inocome_quantile[0.25] and int(row['total_income']) < inocome_quantile[0.5]:
        return 'средий'
    else:
        return "высокий"
bank_data['income_level'] = bank_data.apply(set_income_level, axis = 1 )
bank_data.income_level.value_counts()

высокий    10727
низкий      5364
средий      5363
Name: income_level, dtype: int64

Заменим числовые значения *dept*(наличие задолженности по возврату кредитов) на текстовые  для наглядности

In [28]:
def refund_result(row):
    if row == 0:
        return "Возврат кредита во время"
    else:
        return "Возврат кредита не во время"
bank_data['debt'] = bank_data['debt'].apply(refund_result)

In [29]:
bank_data

children  days_employed  dob_years education  education_id  \
0             1           8437         42    Высшее             0   
1             1           4024         36   Среднее             1   
2             0           5623         33   Среднее             1   
3             3           4124         32   Среднее             1   
4             0         340266         53   Среднее             1   
...         ...            ...        ...       ...           ...   
21520         1           4529         43   Среднее             1   
21521         0         343937         67   Среднее             1   
21522         1           2113         38   Среднее             1   
21523         3           3112         38   Среднее             1   
21524         2           1984         40   Среднее             1   

          family_status  family_status_id gender income_type  \
0       женат / замужем                 0      F   сотрудник   
1       женат / замужем                 0      F   сотрудник   
2       женат / замужем                 0      M   сотрудник   
3       женат / замужем                 0      M   сотрудник   
4      гражданский брак                 1      F   пенсионер   
...                 ...               ...    ...         ...   
21520  гражданский брак                 1      F   компаньон   
21521   женат / замужем                 0      F   пенсионер   
21522  гражданский брак                 1      M   сотрудник   
21523   женат / замужем                 0      M   сотрудник   
21524   женат / замужем                 0      F   сотрудник   

                              debt  total_income  \
0         Возврат кредита во время        253875   
1         Возврат кредита во время        112080   
2         Возврат кредита во время        145885   
3         Возврат кредита во время        267628   
4         Возврат кредита во время        158616   
...                            ...           ...   
21520     Возврат кредита во время        224791   
21521     Возврат кредита во время        155999   
21522  Возврат кредита не во время         89672   
21523  Возврат кредита не во время        244093   
21524     Возврат кредита во время         82047   

                            purpose lemmatized_purpose income_level  
0                     покупка жилья       недвижимость      высокий  
1           приобретение автомобиля         автомобиль       средий  
2                     покупка жилья       недвижимость      высокий  
3        дополнительное образование        образование      высокий  
4                   сыграть свадьбу            свадьба      высокий  
...                             ...                ...          ...  
21520             операции с жильем       недвижимость      высокий  
21521          сделка с автомобилем         автомобиль      высокий  
21522                  недвижимость       недвижимость       средий  
21523  на покупку своего автомобиля         автомобиль      высокий  
21524         на покупку автомобиля         автомобиль       низкий  

[21454 rows x 14 columns]

##### 2.5.2 Определим категории клиентов в зависимости от наличия детей в семье:

In [30]:
def children_category(row):
    """
    Функция для определения категории клиента в
    зависимости от наличия детей в семье
    """
    if row > 0:
        return "есть дети"
    else:
        return "дети отсутствуют"   
bank_data['child'] = bank_data.children.apply(children_category) 
bank_data


children  days_employed  dob_years education  education_id  \
0             1           8437         42    Высшее             0   
1             1           4024         36   Среднее             1   
2             0           5623         33   Среднее             1   
3             3           4124         32   Среднее             1   
4             0         340266         53   Среднее             1   
...         ...            ...        ...       ...           ...   
21520         1           4529         43   Среднее             1   
21521         0         343937         67   Среднее             1   
21522         1           2113         38   Среднее             1   
21523         3           3112         38   Среднее             1   
21524         2           1984         40   Среднее             1   

          family_status  family_status_id gender income_type  \
0       женат / замужем                 0      F   сотрудник   
1       женат / замужем                 0      F   сотрудник   
2       женат / замужем                 0      M   сотрудник   
3       женат / замужем                 0      M   сотрудник   
4      гражданский брак                 1      F   пенсионер   
...                 ...               ...    ...         ...   
21520  гражданский брак                 1      F   компаньон   
21521   женат / замужем                 0      F   пенсионер   
21522  гражданский брак                 1      M   сотрудник   
21523   женат / замужем                 0      M   сотрудник   
21524   женат / замужем                 0      F   сотрудник   

                              debt  total_income  \
0         Возврат кредита во время        253875   
1         Возврат кредита во время        112080   
2         Возврат кредита во время        145885   
3         Возврат кредита во время        267628   
4         Возврат кредита во время        158616   
...                            ...           ...   
21520     Возврат кредита во время        224791   
21521     Возврат кредита во время        155999   
21522  Возврат кредита не во время         89672   
21523  Возврат кредита не во время        244093   
21524     Возврат кредита во время         82047   

                            purpose lemmatized_purpose income_level  \
0                     покупка жилья       недвижимость      высокий   
1           приобретение автомобиля         автомобиль       средий   
2                     покупка жилья       недвижимость      высокий   
3        дополнительное образование        образование      высокий   
4                   сыграть свадьбу            свадьба      высокий   
...                             ...                ...          ...   
21520             операции с жильем       недвижимость      высокий   
21521          сделка с автомобилем         автомобиль      высокий   
21522                  недвижимость       недвижимость       средий   
21523  на покупку своего автомобиля         автомобиль      высокий   
21524         на покупку автомобиля         автомобиль       низкий   

                  child  
0             есть дети  
1             есть дети  
2      дети отсутствуют  
3             есть дети  
4      дети отсутствуют  
...                 ...  
21520         есть дети  
21521  дети отсутствуют  
21522         есть дети  
21523         есть дети  
21524         есть дети  

[21454 rows x 15 columns]

In [31]:
family_dict = bank_data[['family_status','family_status_id']].drop_duplicates().reset_index(drop=True)
education_dict = bank_data[['education','education_id']].drop_duplicates().reset_index(drop=True)
family_dict


family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  Не женат / не замужем                 4

In [32]:
education_dict

education  education_id
0               Высшее             0
1              Среднее             1
2  Неоконченное высшее             2
3            Начальное             3
4       Ученая степень             4

### 3. Определение зависимостей различных факторов на возврат кредита в срок

#### 3.1 Опредление влияния наличия детей на возврат кредита в срок

Для определения влияния построим сводную таблицу:

In [33]:
pivot_data_children = bank_data.pivot_table(index=['child'], values = 'children', columns='debt', aggfunc='count',fill_value = 0)
pivot_data_children['Доля возврата не вовремя'] = round(pivot_data_children['Возврат кредита не во время']/pivot_data_children['Возврат кредита во время'], 3)
pivot_data_children = pivot_data_children.sort_values(by = 'Доля возврата не вовремя')
pivot_data_children

debt              Возврат кредита во время  Возврат кредита не во время  \
child                                                                     
дети отсутствуют                     13142                         1072   
есть дети                             6571                          669   

debt              Доля возврата не вовремя  
child                                       
дети отсутствуют                     0.082  
есть дети                            0.102

**Вывод:**

Из полученных данных видно, что доля клиентов, которые возвращают кредит не во время меньше, чем доля клиентов с детьми

##### 3.2 Опредление влияния семейного положения на возврат кредита в срок

In [34]:
pivot_data_family = bank_data.pivot_table(index=['family_status'], values = 'education', columns='debt', aggfunc='count',fill_value = 0)
pivot_data_family['Доля возврата не вовремя'] = round(pivot_data_family['Возврат кредита не во время']/pivot_data_family['Возврат кредита во время'], 3)

In [35]:
pivot_data_family.sort_values(by = 'Доля возврата не вовремя')

debt                   Возврат кредита во время  Возврат кредита не во время  \
family_status                                                                  
вдовец / вдова                              896                           63   
в разводе                                  1110                           85   
женат / замужем                           11408                          931   
гражданский брак                           3763                          388   
Не женат / не замужем                      2536                          274   

debt                   Доля возврата не вовремя  
family_status                                    
вдовец / вдова                            0.070  
в разводе                                 0.077  
женат / замужем                           0.082  
гражданский брак                          0.103  
Не женат / не замужем                     0.108

**Вывод:**

Из полученных данных можем сделать вывод, что наиболее надежные заемщики, это те, кто находится или когда-либо был в официальном браке.

##### 3.3 Опредление влияния уровня доходов на возврат кредита в срок

In [36]:
pivot_data_income = bank_data.pivot_table(index=['income_level'], values = 'total_income', columns='debt', aggfunc='count',fill_value = 0)
pivot_data_income['Доля возврата не вовремя'] = round(pivot_data_income['Возврат кредита не во время']/pivot_data_income['Возврат кредита во время'], 3)
pivot_data_income = pivot_data_income.sort_values(by = 'Доля возврата не вовремя')
pivot_data_income

debt          Возврат кредита во время  Возврат кредита не во время  \
income_level                                                          
низкий                            4945                          419   
высокий                           9863                          864   
средий                            4905                          458   

debt          Доля возврата не вовремя  
income_level                            
низкий                           0.085  
высокий                          0.088  
средий                           0.093

**Вывод:**

Из полученных данных следует, что  люди с высоким уровнем дохода реже возвращают кредит не вовремя чем люди со средним и низким уровнем дохода. Чаще других кредит возвращают не во время люди со средним уровнем дохода.

##### 3.4 Опредление влияния цели кредита на возврат кредита в срок

In [37]:
pivot_data_purpose = bank_data.pivot_table(index=['lemmatized_purpose'], values = 'total_income', columns='debt', aggfunc='count',fill_value = 0)
pivot_data_purpose['Доля возврата не вовремя'] = round(pivot_data_purpose['Возврат кредита не во время']/pivot_data_purpose['Возврат кредита во время'], 3)
pivot_data_purpose = pivot_data_purpose.sort_values(by = 'Доля возврата не вовремя')
pivot_data_purpose

debt                Возврат кредита во время  Возврат кредита не во время  \
lemmatized_purpose                                                          
недвижимость                           10029                          782   
свадьба                                 2138                          186   
образование                             3643                          370   
автомобиль                              3469                          359   

debt                Доля возврата не вовремя  
lemmatized_purpose                            
недвижимость                           0.078  
свадьба                                0.087  
образование                            0.102  
автомобиль                             0.103

**Вывод:**

Видно, что клиенты чаще во время возвращают кредитны на недвижимость, чуть реже на свадьбу и еще реже на автомобиль и обазование.

### 4 Вывод

На основании проведенной работы и полученных результатов, что клиент, который с большей долей вероятности вернет кредит вовремя - это клиент с высоким уровнем дохода, который женат или замужем, без детей и он берет кредит на недвижимость.

А клиент, который с наименьшей долей вероятности вернет кредит вовремя - это не женатый человек со средним образованием у которого есь дети, и кредит он берет на автомобиль или образование. 